spec/index.md

```diff
@@ -4620,14 +4679,23 @@ S.system_time <= Q.ingress_expiry
 
 ```
 
-Query response  
+Query response `R`:
+
 -   if `composite_query_helper(S, MAX_CYCLES_PER_QUERY, 0, Q.canister_id, Q.sender, Q.canister_id, Q.method_name, Q.arg) = (Reject (RejectCode, RejectMsg), _)` then
 
-        {status: "rejected"; reject_code: RejectCode; reject_message: RejectMsg; error_code: <implementation-specific>}
+        {status: "rejected"; reject_code: RejectCode; reject_message: RejectMsg; error_code: <implementation-specific>, signatures: Sigs}
+
+-   Else if `composite_query_helper(S, MAX_CYCLES_PER_QUERY, 0, Q.canister_id, Q.sender, Q.canister_id, Q.method_name, Q.arg) = (Reply Res, _)` then
+
+        {status: "replied"; reply: {arg: Res}, signatures: Sigs}
 
--   Else if `composite_query_helper(S, MAX_CYCLES_PER_QUERY, 0, Q.canister_id, Q.sender, Q.canister_id, Q.method_name, Q.arg) = (Reply R, _)` then
+where the query `Q`, the response `R`, and a certificate `Cert'` that is obtained by requesting the path `/subnet` in a **separate** read state request to `/api/v2/canister/<effective_canister_id>/read_state` satisfy the following:
```

Comment:
> requesting the path `/subnet`

In this way, users don't need to specify the subnet_id, which they may not know in advance.

However, in implementation, the `public_key` and `canister_ranges` for all subnets are populated in the state tree in each subnet. As a result, in addition to the node keys, uses get many extra paths, resulting a larger certificate size.

Suppose we have 36 subnets, and canister ranges only have one entry of (principal, principal). Each principal is 29 bytes and the subnet public key is 48 bytes. In total, that's 36 * (29 + 29 + 48) = 3816 bytes.

I had some early experiments showing that with 40 node keys, the tree size is 2800 bytes. I need to do some more experiments on the tree size with paths of all subnets. I estimate it will be around 3 times larger but not that large to make a significant difference on network transmission.

Still, it would be great if there is a simple way to get rid paths of other subnets in this read_state call.

Reply:
Thanks a lot for pointing this out. This is very useful data. We discussed multiple ways of working around this in the future. Options could, e.g., be

- Providing an API that allows clients to obtain the subnet ID that hosts a canister based on its ID. 
- Allowing to request the certificate as part of the response.

For this iteration I'd consider solving this out of scope, though.

Reply:
The [verify_responses()](https://github.com/dfinity/interface-spec/pull/163/files#diff-b9a833d6114611a080aee4cd57618feb45aa044bcc11105d44b155b5bbde00ebR759) method above checks:
 `Cert.delegation = NoDelegation ∧ SubnetId = RootSubnetId` or `SubnetId = Cert.delegation.subnet_id`. However, users do not know the subnet id in the first place. 

Actually, if users do not know the subnet id in advance, the condition above serves as the way to figure out the subnet id after having the certificate. @mraszyk

Reply:
From the workflow above, there are some awkward part related to subnet id:
a. Before the read_state call, users do not know the subnet_id and thus request the whole `/subnet` path
b. After getting the certificate, users get the subnet_id from the delegation. If no delegation, it implies the subnet is root subnet but there is no hardcoded root subnet id. Hardcoded root subnet id also brings problems in testnet settings.
c.  With the whole tree at path `/subnet`, users need to know the subnet id to look up the node keys. Otherwise, it has to loop on all forks until it find the fork with node keys.

In summary, there are some options before request and after response:
Before request:
1. Providing an API that allows clients to obtain the subnet ID that hosts a canister based on its ID.
2. Allow users requesting node keys with a default subnet id
3. Considering not put node keys under `/subnet/<subnet-id>/node` but directly under **/metadata/node**. cc: @mraszyk We used to argue for `/subnet/<subnet-id>/node` because it conveys the subnet membership. However, it seems the membership information is anyway not straightforward if not knowing the subnet id.
 
After response:
1. Add /metadata/own_subnet_id in the state tree and include it in the certificate. So users has an easy way to know about subnet_id if delegation is not present.

Reply:
Thanks @ShuoWangNSL for your explanations. I think we should not overcomplicate this. The way to get the subnet ID you describe above is IMO good enough for the initial iteration (and what we had agreed upon IIRC). We can always improve moving forward.

Reply:
How about returning the subnet ID as an additional field in the query response?

Reply:
Regarding option 3, putting node keys under `/subnet/<subnet-id>/node` gives us the benefit of being able to look at the state tree as one global tree. Therefore, let's keep this one.

The options 1 , 2, and 4 can be considered later. We can use the current way of obtaining subnet_id in our initial iteration.

Reply:
> How about returning the subnet ID as an additional field in the query response?

As the read_state is a separate call, I think it is not always feasible to use the information from another call. In practice, I think agents may want parallelize these two calls to reduce latency.

Reply:
> Regarding option 3, putting node keys under /subnet/<subnet-id>/node gives us the benefit of being able to look at the state tree as one global tree. Therefore, let's keep this path.

Agreed!

> In practice, I think agents may want to parallelize these two calls to reduce latency.

Good point! Should we then close this thead?